### PR TITLE
Test n+1

### PR DESCRIPTION
**Pull request by PierreBarba**
# hello reviewer I hope you are having a great day, any suggestion will be welcome, thank you very much for the support 😄 🌴

## Ruby requirements

> Follow our list of [best practices for Ruby](https://github.com/microverseinc/curriculum-ruby/blob/main/articles/ruby_best_practices.md).

### n+1

- Make sure that the N+1 problem is solved when fetching all posts and their comments for a user.

> In your PR description include the screenshots of your app's logs before and after the fix

Captura de pantalla 2023-04-16 a la(s) 17 26 04

Captura de pantalla 2023-04-16 a la(s) 17 26 18

> Do not use the bullet gem in this exercise - instead analyze the logs as showed in this [article](https://dev.to/junko911/rails-n-1-queries-and-eager-loading-10eh)

- This is the corresponding wireframe view: blog all user posts

- Integration specs
- Use Capybara to write integration tests for each view in your project.

> Think on the flow of things and group the tests of page in a way that makes sense. For example, one test would check for the things in the page and another test for clicking links on it.

### User index page:

- I can see the username of all other users.
- I can see the profile picture for each user.
- I can see the number of posts each user has written.
- When I click on a user, I am redirected to that user's show page.

### user show page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see the user's bio.
- I can see the user's first 3 posts.
- I can see a button that lets me view all of a user's posts.
- When I click a user's post, it redirects me to that post's show page.
- When I click to see all posts, it redirects me to the user's post's index page.

### User post index page:

- I can see the user's profile picture.
- I can see the user's username.
- I can see the number of posts the user has written.
- I can see a post's title.
- I can see some of the post's body.
- I can see the first comments on a post.
- I can see how many comments a post has.
- I can see how many likes a post has.
- I can see a section for pagination if there are more posts than fit on the view.
- When I click on a post, it redirects me to that post's show page.

### Post show page:

- I can see the post's title.
- I can see who wrote the post.
- I can see how many comments it has.
- I can see how many likes it has.
- I can see the post body.
- I can see the username of each commentor.
- I can see the comment each commentor left.

### CODE REVIEW

File: spec/views/users_show_spec.rb [hunk 1]
```diff
@@ -0,0 +1,50 @@
+require 'rails_helper'
+
+RSpec.describe 'User show page - ', type: :feature do
```
Comment (victorokoroji) [inline spec/views/users_show_spec.rb:3]: - [ ] Good job so far with your tests 👍🏽 . When you click on a user's post, kindly add a test that redirects to the post's show page as specified in the project requirement

File: spec/views/posts_index_spec.rb [hunk 2]
```diff
@@ -0,0 +1,64 @@
+require 'rails_helper'
+
+RSpec.describe 'post_index', type: :feature do
```
Comment (victorokoroji) [inline spec/views/posts_index_spec.rb:3]: - [ ] Good job so far with your tests 👍🏽 .Kindly add tests that assert you can see how many likes a post has as specified in the project requirement

File: spec/controllers/posts_controller_spec.rb [hunk 3]
```diff
@@ -1,36 +1,45 @@
...
+      photo: 'https://unsplash.com/photos/Hekapoo',
+      bio: 'User 1 bio',
+      posts_counter: 0
+    )
+  end
+
   describe 'GET /index' do
     it 'returns http success' do
-      get '/users/:id/posts/'
+      get users_path
```
Comment (victorokoroji) [inline spec/controllers/posts_controller_spec.rb:14]: [OPTIONAL]

you have a failed test in your controller. You might want to check it out 😄 

image

**Review by victorokoroji — CHANGES_REQUESTED** [review]
Hi @PierreBarba, 

You have done a great job so far! 💪 👏
You still need to work on a few issues before moving on to the next project, but you're almost there!

## Top highlight 🎯

- The Pull Request is following the Git Flow ✅ 
- All feature tests are passing ✅ 
- The PR title and description are professional and descriptive ✅ 
-Tests are well written ✅ 
- The README is well documented ✅ 
- Linters are working properly ✅ 


## Required Changes ♻️
_Check the comments under the review._


## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:

N/A

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users_show_spec.rb [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+  it "When I click to see posts, it redirects me to the user's post's index page." do
+    click_link 'See post'
+    expect(page).to have_current_path(user_posts_path(@user1))
+  end
+
+  it 'When I click a users post, it redirects me to that posts show page.' do
+    link_to @post1.text
+    expect(page).to have_current_path(user_post_path(@user1, @post1))
+  end
```
Comment (TedLivist) [inline spec/views/users_show_spec.rb]: - [ ] This test is failing. Kindly fix that.

  image

**Review by TedLivist — CHANGES_REQUESTED** [review]
## STATUS: CHANGES REQUIRED ♻️ ♻️

Hi @PierreBarba and @ivanmvh,

✔️ Scenarios are well covered

Great work on making the changes requested by a previous reviewer 👏 
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.

 ## Required Changes ♻️
_Check the comments under the review._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear. And please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: spec/views/users_show_spec.rb (hunk 4 above)
Comment (PierreBarba) [reply spec/views/users_show_spec.rb]: Hello, sorry to bother you. I'm running the tests and they pass on my computer. Do you think we could review it?

Comment (TedLivist) [reply spec/views/users_show_spec.rb]: Hi @PierreBarba sure

Comment (TedLivist) [reply spec/views/users_show_spec.rb]: Call me on Zoom

**Review by brytebee — APPROVED** [review]
Hi, @PierreBarba 👋,

##

### Highlights💡
- Good commit message.✔️ 
- Good PR Summary.✔️
- Great UI design.✔️
- App functions as it should.✔️

##

# STATUS: APPROVED 🕺🥳 

Congratulations🍾🥳!! Your project is ready for presentation!  There is nothing else to say other than... it's time to merge it :shipit: 



Good luck! 🎉 

##

###  **Optional Suggestions:**
- N/A

------

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me @brytebee in your question so I can receive the notification.
### _**Use @brytebee on slack for faster responses.**_

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**